### PR TITLE
*: add openshift-etcd quorum-guard

### PR DESCRIPTION
**Pull request by hexfusion**
Testing adding basic quorum guard to `openshift-etcd` namespace. Next steps for this PR will be to actively remove the CVO assets from MCO namespace so we can remove quorum-guard in MCO and merge CI.

Signed-off-by: Sam Batschelet <sbatsche@redhat.com>

### CODE REVIEW

File: pkg/operator/guardbudget/guard_budget_controller.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+	// check that it exists in our namespace
+	c.eventRecorder.Event("GuardBudgetController", "safe to remove pod disruption budget in MCO namespace")
+	// remove here
+
+	return nil
+}
+
+func (c *GuardBudgetController) budgetLister(namespace, name string) (bool, error) {
+	machineConfigOperatorPDB, err := c.machineConfigOperatorPodDisruptionBudgetLister.PodDisruptionBudgets(namespace).Get(name)
```
Comment (hexfusion) [inline pkg/operator/guardbudget/guard_budget_controller.go]: not sure how this will work generically.....

File: pkg/operator/starter.go [hunk 2]
```diff
@@ -66,6 +67,7 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 		operatorclient.OperatorNamespace,
 		"openshift-kube-apiserver",
 		"openshift-etcd",
+		"machine-config-operator", // TODO remove after quorum-guard is removed from MCO
```
Comment (hexfusion) [inline pkg/operator/starter.go]: typo openshift-..

File: pkg/operator/guardbudget/guard_budget_controller.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+			operatorClient.Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor("openshift-machine-config-operator").Policy().V1beta1().PodDisruptionBudgets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor("openshift-etcd").Policy().V1beta1().PodDisruptionBudgets().Informer().HasSynced,
+		},
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "GuardBudgetController"),
+		eventRecorder: eventRecorder.WithComponentSuffix("guard-budget-controller"),
+	}
+
+	kubeInformersForNamespaces.InformersFor("openshift-machine-config-operator").Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForNamespaces.InformersFor("openshift-etcd").Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
```
Comment (hexfusion) [inline pkg/operator/guardbudget/guard_budget_controller.go]: remove

**Comment by hexfusion** [comment]
/hold

manual testing on top of 142

File: manifests/0000_12_etcd-operator_00_pdb-rolebinding.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: openshift-machine-config-operator
+  name: pdb-mco
+rules:
+- apiGroups: ["policy"]
+  resources: ["poddisruptionbudgets"]
+  verbs: ["get", "watch", "list", "delete"]
```
Comment (deads2k) [inline manifests/0000_12_etcd-operator_00_pdb-rolebinding.yaml]: no create or update?

Comment (deads2k) [reply manifests/0000_12_etcd-operator_00_pdb-rolebinding.yaml]: > no create or update?

oh, I see, this is for deletion powers.  ok

File: manifests/0000_12_etcd-operator_09_etcdquorumguard_deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      - image: quay.io/openshift/origin-cli:latest
+        imagePullPolicy: IfNotPresent
+        name: guard
+        terminationMessagePolicy: FallbackToLogsOnError
+        volumeMounts:
+        - mountPath: /var/run/secrets/etcd-client
+          name: etcd-client
+        - mountPath: /var/run/configmaps/etcd-serving-ca
+          name: etcd-serving-ca
+        command:
```
Comment (deads2k) [inline manifests/0000_12_etcd-operator_09_etcdquorumguard_deployment.yaml]: all copy/paste?

File: pkg/operator/guardbudget/guard_budget_controller.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *GuardBudgetController) checkGuardBudget() error {
+	// checks if quorum-guard pdb exists in mco namespace
+	machineConfigOperatorPDB, err := c.machineConfigOperatorPodDisruptionBudgetLister.PodDisruptionBudgets("openshift-machine-config-operator").Get("etcd-quorum-guard")
+	if err != nil {
```
Comment (deads2k) [inline pkg/operator/guardbudget/guard_budget_controller.go]:
```
if isnotFound(err){
}
if err != nil{
}
```

to reduce nesting

File: pkg/operator/guardbudget/guard_budget_controller.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		Status:  operatorv1.ConditionFalse,
+		Reason:  "GuardBudgetNotRemoved",
+		Message: fmt.Sprintf("Quorum guard pod disruption budget is not removed yet: machineConfigOperatorPDB %#v", machineConfigOperatorPDB),
+	}))
+
+	if err := c.kubeClient.PolicyV1beta1().PodDisruptionBudgets("openshift-machine-config-operator").Delete("etcd-quorum-guard", nil); err != nil {
+		c.eventRecorder.Event("GuardBudgetController", "failed to remove guard budget")
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
```
Comment (deads2k) [inline pkg/operator/guardbudget/guard_budget_controller.go]: no, you'll get triggered again because of the delete. Let the re-enter on this loop set it.

**Comment by deads2k** [comment]
nothing fatal, but worth updating
/lgtm

**Comment by hexfusion** [comment]
David picked this onto #142

**Pull request closed without merging**